### DISCUSSION TITLE
Recommendation for feature flag name that enables 'serde' and also enables 'serde' in dependencies

### DISCUSSION THREAD

**Discussion by frewsxcv** [rust-lang-nursery/api-guidelines#180] [General]
> The implicit feature inferred by Cargo for the optional dependency is called `serde`, not `use-serde` or `with-serde`, so we like for explicit features to behave the same way.

https://rust-lang-nursery.github.io/api-guidelines/naming.html#feature-names-are-free-of-placeholder-words-c-feature

What is the recommended naming convention for a serde feature flag that also enables serde feature flags of dependencies? As an example:

```toml
[features]
use-serde = ["serde", "geo-types/serde"]
```

- https://github.com/georust/geo/blob/b57ab9ffac65e0ce45e549c45acc4ef6408fd310/geo/Cargo.toml#L29

**Comment by dtolnay** [comment]
Ideally it would still be `"serde"`. This requires the `namespaced-features` feature of Cargo: https://github.com/rust-lang/cargo/issues/5565#issuecomment-412296421.

Hmm, on stable Cargo it may be possible to work around it with a shim crate that re-exports from serde:

```ini
[features]
serde = ["serde1", "geo-types/serde"]

[dependencies]
geo-types = "..."
serde1 = { version = "1.0", optional = true }
```

I don't currently have a suggestion for a name other than `"serde"`. :grinning:

**Comment by BurntSushi** [comment]
I've seen `serde1` used somewhere as the feature name, although I don't remember where.

**Comment by KodrAus** [comment]
@BurntSushi [`rand` is one example that uses `serde1`](https://github.com/rust-random/rand/blob/master/Cargo.toml#L28).

**Comment by tikue** [comment]
@dtolnay as I found out the hard way, crates that use `namespaced-features` can't currently be published.

**Comment by grovesNL** [comment]
If anyone else is looking for a way to just expose the feature as `serde` in this case (avoiding having to name the feature something like `serde1` or `use-serde`), the approach in https://github.com/RustCrypto/RSA/pull/41/files seems to be a good workaround and seems to be used in a few crates now.

**Comment by CAD97** [comment]
Rather than use `#[serde(serde_crate)]`, consider an `extern crate serde_crate as serde;` at the top of the library. This reexports the library _back_ as just `::serde`.

**Comment by mainrs** [comment]
@CAD97 in the library that exposes a feature called `serde`? I've added `#[cfg(feature = "serde")]
extern crate serde_crate as serde;` but I still get the compilation error: 

```rust
can't find crate for `serde`
 --> zors_lib\src\visitable\mod.rs:9:38
  |
9 | #[cfg_attr(feature = "serde", derive(Deserialize))]
  |                                      ^^^^^^^^^^^ can't find crate
```

**Comment by CAD97** [comment]
... I could have sworn that I did this before and it worked. (It must have been for a project that was not using the derive, or with a macro that wasn't actually serde's.)

But yes, now that I think about it, serde_derive does `extern crate serde` (and in `edition2015` mode), so wouldn't work with `extern crate` re-renaming. (If `serde_derive` were to change to using `edition2018`, it would probably be better to emit `use ::serde`, which unambiguously means an extern crate in `edition2018` mode.)

And this gets worse on nightly, because serde is in the sysroot 😞 

```
D:\repos\cad97\playground>cargo +beta check --quiet
error[E0463]: can't find crate for `serde`
 --> src\lib.rs:5:21
  |
5 | #[derive(Serialize, Deserialize)]
  |                     ^^^^^^^^^^^ can't find crate
  |
  = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to previous error

error: could not compile `playground`.

D:\repos\cad97\playground>cargo +nightly check --quiet
error[E0658]: use of unstable library feature 'rustc_private': this crate is being loaded from the sysroot, an unstable location; did you mean to load this crate from crates.io via `Cargo.toml` instead?
 --> src\lib.rs:5:10
  |
5 | #[derive(Serialize)]
  |          ^^^^^^^^^
  |
  = note: see issue #27812 <https://github.com/rust-lang/rust/issues/27812> for more information
  = help: add `#![feature(rustc_private)]` to the crate attributes to enable
  = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to previous error

error: could not compile `playground`.
```

**Reply by bminixhofer** [reply]
For anyone else finding this, there is a solution for the `#[derive(Serialize)]` issue here: https://github.com/serde-rs/serde/issues/1938#issuecomment-751291607

> `#[serde(crate = "serde_by_some_other_name")]` should work.
